File: ios/GradeThread/Disclosure/DisclosureGeometry.swift
import UIKit

/// Pure layout + color math for compositing an annotated defect photo. Kept
/// side-effect-free so the geometry is unit-tested without rendering pixels;
/// `DisclosureRenderer` consumes it. Mirrors the web `annotated-photo.tsx`.
enum DisclosureGeometry {
    /// Cap the composite's long edge so the legend text stays legible and the
    /// PNG stays small. Matches the web `MAX_CANVAS_W`.
    static let maxCanvasWidth: CGFloat = 900
    static let legendLineHeight: CGFloat = 26
    static let legendPadding: CGFloat = 14

    /// Photo draw size: scaled down to `maxCanvasWidth` (never up), aspect kept.
    static func canvasSize(imageSize: CGSize) -> CGSize {
        guard imageSize.width > 0, imageSize.height > 0 else { return .zero }
        let width = min(imageSize.width, maxCanvasWidth)
        let scale = width / imageSize.width
        return CGSize(width: width, height: (imageSize.height * scale).rounded(.down))
    }

    /// Map a normalized `[x, y, w, h]` bbox into pixel space for `size`.
    /// Returns nil for a malformed bbox (caller treats it as legend-only).
    static func scaledRect(bbox: [Double], in size: CGSize) -> CGRect? {
        guard bbox.count == 4 else { return nil }
        return CGRect(
            x: CGFloat(bbox[0]) * size.width,
            y: CGFloat(bbox[1]) * size.height,
            width: CGFloat(bbox[2]) * size.width,
            height: CGFloat(bbox[3]) * size.height
        )
    }

    /// Height of the legend strip beneath the photo for `count` callouts.
    static func legendHeight(annotationCount count: Int) -> CGFloat {
        guard count > 0 else { return 0 }
        return legendPadding * 2 + CGFloat(count) * legendLineHeight
    }

    /// Total composite size = photo + legend strip.
    static func compositeSize(imageSize: CGSize, annotationCount: Int) -> CGSize {
        let canvas = canvasSize(imageSize: imageSize)
        return CGSize(width: canvas.width, height: canvas.height + legendHeight(annotationCount: annotationCount))
    }
}

/// Severity → color, matching the web `SEVERITY_COLOR` (Vibrant Crimson / Amber
/// Gold / Yellow). Unknown severities fall back to the minor tone.
enum DisclosureSeverityColor {
    static func color(for severity: String) -> UIColor {
        switch severity.lowercased() {
        case "major":    return UIColor(red: 0.941, green: 0.239, blue: 0.373, alpha: 1) // #F03D5F
        case "moderate": return UIColor(red: 0.961, green: 0.620, blue: 0.043, alpha: 1) // #F59E0B
        default:         return UIColor(red: 0.918, green: 0.702, blue: 0.031, alpha: 1) // #EAB308
        }
    }
}

File: android/app/src/main/java/com/gradethread/app/disclosure/DisclosureGeometry.kt
package com.gradethread.app.disclosure

import kotlin.math.floor

/**
 * US-1360: where a defect callout lands on the photo.
 *
 * Pure, and mirrors the web `annotated-photo.tsx` constant for constant. That
 * shared arithmetic IS the feature: annotations are stored normalised, so a box
 * drawn over the preview on a phone and the same box burned into the saved PNG
 * (and the one the web renders) must land on the same thread. Any divergence
 * here puts a marker next to a flaw instead of on it, in a photo that goes to a
 * buyer as evidence.
 */
object DisclosureGeometry {

    /** The composite's long edge. Bigger doesn't help; it just bloats the PNG. */
    const val MAX_CANVAS_WIDTH = 900f

    const val LEGEND_LINE_HEIGHT = 26f
    const val LEGEND_PADDING = 14f

    data class Size(val width: Float, val height: Float) {
        val isEmpty: Boolean get() = width <= 0f || height <= 0f
    }

    data class Rect(val left: Float, val top: Float, val width: Float, val height: Float) {
        val right: Float get() = left + width
        val bottom: Float get() = top + height
    }

    /**
     * The photo's drawn size: scaled DOWN to the cap, never up, aspect kept.
     *
     * Never up because upscaling a small flaw photo makes a blurrier image, not
     * a clearer one — and this photo's whole job is showing a buyer the flaw.
     */
    fun canvasSize(imageWidth: Int, imageHeight: Int): Size {
        if (imageWidth <= 0 || imageHeight <= 0) return Size(0f, 0f)
        val width = minOf(imageWidth.toFloat(), MAX_CANVAS_WIDTH)
        val scale = width / imageWidth
        return Size(width, floor(imageHeight * scale))
    }

    /**
     * A normalised `[x, y, w, h]` box in pixels, or null when malformed.
     *
     * Null rather than a guessed rectangle: a callout drawn from bad
     * coordinates points at the wrong part of the garment, which is worse than
     * no box at all — the legend still names the defect.
     */
    fun scaledRect(bbox: List<Double>?, size: Size): Rect? {
        if (bbox == null || bbox.size != 4) return null
        if (size.isEmpty) return null
        if (bbox.any { !it.isFinite() }) return null
        return Rect(
            left = (bbox[0] * size.width).toFloat(),
            top = (bbox[1] * size.height).toFloat(),
            width = (bbox[2] * size.width).toFloat(),
            height = (bbox[3] * size.height).toFloat(),
        )
    }

    /** The legend strip under the photo. Zero when there's nothing to list. */
    fun legendHeight(annotationCount: Int): Float =
        if (annotationCount <= 0) 0f else LEGEND_PADDING * 2 + annotationCount * LEGEND_LINE_HEIGHT

    /** Photo plus legend. */
    fun compositeSize(imageWidth: Int, imageHeight: Int, annotationCount: Int): Size {
        val canvas = canvasSize(imageWidth, imageHeight)
        return Size(canvas.width, canvas.height + legendHeight(annotationCount))
    }

    /**
     * Severity → colour, matching the web's SEVERITY_COLOR.
     *
     * ⚠ US-3010: these stay raw and fixed. DisclosureRenderer paints them into
     * a BITMAP that is exported and shown to buyers, so they must match the
     * web's values rather than the reader's theme - the same annotation has to
     * be the same colour wherever the image is opened.
     */
    object SeverityColor {
        const val MAJOR = 0xFFF03D5F.toInt()
        const val MODERATE = 0xFFF59E0B.toInt()
        const val MINOR = 0xFFEAB308.toInt()

        /** Unknown severities fall back to the minor tone rather than vanishing. */
        fun of(severity: String): Int = when (severity.lowercase()) {
            "major" -> MAJOR
            "moderate" -> MODERATE
            else -> MINOR
        }
    }

    /** "3. Pilling on the left cuff (minor)" — the legend line for a callout. */
    fun legendLine(annotation: PhotoAnnotation): String {
        val where = annotation.location?.takeIf { it.isNotBlank() }?.let { " — $it" }.orEmpty()
        val severity = annotation.severity.takeIf { it.isNotBlank() }?.let { " ($it)" }.orEmpty()
        return "${annotation.n}. ${annotation.issue}$where$severity"
    }
}
